### ISSUE TITLE
Program stops working when EMTERPRETIFY enabled

### ISSUE THREAD

**Issue by Cyrille-de-Brebisson** [emscripten-core/emscripten#6684]
Hello,

I have a fairly large project that I am trying to port to web-assembly.

I am in the early stages at this point and I wanted to use the EMTERPRETIFY option in order to enable some of the async features.

However, while my program does "boot" in "normal" mode, when I enable EMTERPRETIFY, I get a crash before init time.

- My application now throws an exception at start time: "Exception thrown, see JavaScript console" with:
  
RangeError: invalid or out-of-range index prime.js:584:7
allocate    file:///P:/trv/fir/FIR/_Webasm/build/prime.js:584:7
<anonyme>  file:///P:/trv/fir/FIR/_Webasm/build/prime.js:8000:22

with line 8000 being: var ___tm_timezone=allocate(intArrayFromString("GMT"), "i8", ALLOC_STATIC);
and line 584 being part of the allocate function
  if (singleType === 'i8') {
    if (slab.subarray || slab.slice) {
      HEAPU8.set(/** @type {!Uint8Array} */ (slab), ret); //<<< line 584
    } else {
      HEAPU8.set(new Uint8Array(slab), ret);
    }
    return ret;
  }

prime.js is of course 100% automatically generated by emcc...

I can provide the binary if needed (it is around 18MB!)...

Cyrille

**Comment by kripken** [comment]
Adding some logging for the length of HEAPU8 and the value of `ret` there might help. Presumably `ret` is too large or something like that.

One possibility is that the emterpreter mode uses more memory - it stores the bytecode in memory. Perhaps it simply runs out. If so, increasing  TOTAL_MEMORY might fix this. However, this is still a bug if so, we should be reporting a proper error in such a case.

**Comment by Cyrille-de-Brebisson** [comment]
Hello,

Sorry, but I am a total Newbie here (never done any JS or html before!)... Although I am an experienced C programmer... So I understand the geist of what you are suggesting, but I have no clue how to implement it.

printf in the allocate function does not seem to work. Can you give me the exact line of code to add there to display (in the console I assume), the HEAPU8 size, and also how can I update the TOTAL_MEMORY (I compile with ALLOW_MEMORY_GROWTH=1 BTW).

Cyrille

**Comment by kripken** [comment]
Sure, something like this will print out hopefully interesting stuff:
```
console.log(['debugging', HEAPU8.length, ret]);
```
You should not need to modify TOTAL_MEMORY, it should be handled automatically. If that logging confirms `ret` is too large, then my guess is that when we allocate memory for the emterpreter bytecode we don't handle growth there, and a growth was needed.

**Comment by Cyrille-de-Brebisson** [comment]
Hello,

Adding the log in the allocate function produces this output:

Array(3) [ "debugging", 16777216, 28447808 ] prime.js:582:1
RangeError: invalid or out-of-range index 
allocate prime.js:586:7
<anonyme> file:///P:/trv/fir/FIR/_Webasm/build/prime.js:586:7

I am assuming that the 16MB is the heap size, but I do not know what the 28M relates to... My app binary file is in the 17MB range BTW...

Cyrille

**Comment by kripken** [comment]
Thanks, that confirms the theory!

With that info I managed to created a testcase I can debug locally, using
```
./emcc tests/hello_libcxx.cpp -s LINKABLE=1 -s EMTERPRETIFY=1 -s EMTERPRETIFY_FILE=a.dat -o a.html -s TOTAL_MEMORY=1MB -s TOTAL_STACK=5KB
```
And I see the exact same error as you. Debugging it, the issue was a lack of an assertion on having enough room for static allocations. Fix is in #6721. For your app, you need to build with a larger value of TOTAL_MEMORY.

**Comment by Cyrille-de-Brebisson** [comment]
Hello,

I bumped the memory to 64M, and it now works! Thanks for finding the problem.

However, it is very slow (order of magnitude slower than the "normal" version)...
Is this normal/expected?

Cyrille

**Comment by Cyrille-de-Brebisson** [comment]
Hello,

I am not sure that I understood the whitelist stuff properly.
As far as I can tell, it allows me to tell the compiler/linker that a given function should be not interpreted...

However, my problem is kind of reverse here...
Basically, I have a "classic" Read Eval Print loop. and I can break said loop and use the normal, non interpreted version of scripts...
Except that, in some (rare cases), the Read or Eval might lead to a recursive call to the REPL loop (good example are message boxes)...
So, I have a list of code that NEEDS to be interpreted, and ALL the rest should NOT be interpreted...
Is there a way to handle this?

Basically, what I need to be able to do is to run my code "normally", in a non interpreted way, but jump to interpreted when a given function is called (typically the REPL loop itself)...

Is there a way to do something like this?

Cyrille

**Comment by Cyrille-de-Brebisson** [comment]
Hello,

Thanks for the info... Unfortunately, I can not rework my program this way... Mostly because my program has a programming language of it's own that the user can use, and that programming language is what/where recursive REPL can happen...

User written functions can override various part of the system and be called both during the read, the eval and the print part of the REPLoop :-(

Cyrille

**Comment by curiousdannii** [comment]
In that situation most of your program will probably have to be interpreted then, with just isolated stand alone functions which don't call back into the REPL loop being run normally. Which unfortunately will be slow.
